src/Kafka/Consumer/Callbacks.hs

```diff
@@ -54,19 +58,24 @@ redirectPartitionQueue (Kafka k) (TopicName t) (PartitionId p) q = do
     Nothing -> return ()
     Just pq -> rdKafkaQueueForward pq q
 
-setRebalanceCallback :: (KafkaConsumer -> RebalanceEvent -> IO ())
+setRebalanceCallback :: (KafkaConsumer -> RebalanceEvent -> IO (Maybe [TopicPartition]))
                           -> KafkaConsumer
                           -> KafkaError
                           -> RdKafkaTopicPartitionListTPtr -> IO ()
 setRebalanceCallback f k e pls = do
   ps <- fromNativeTopicPartitionList'' pls
-  let assignment = (tpTopicName &&& tpPartition) <$> ps
   let (Kafka kptr) = getKafka k
 
   case e of
     KafkaResponseError RdKafkaRespErrAssignPartitions -> do
-        f k (RebalanceBeforeAssign assignment)
-        void $ rdKafkaAssign kptr pls
+        -- Consumer may want to alter the offsets if they are stored locally.
+        mbAltered <- f k (RebalanceBeforeAssign ps)
+        (pls', assigned) <- case mbAltered of
+          Nothing -> pure (pls, ps)
+          Just alt -> do
+            pls' <- toNativeTopicPartitionList alt
+            pure (pls', alt)
```

Comment:
Can we add some code to guarantee that all (and only) the partitions from the "original" assignment are passed to `rdKafkaAssign`?

Currently there is a possibility for me to return `Just` a subset (or a superset) of the partitions. Which may cause problems.
But if we take a list of `altered` partitions, add the "missing" ones and remove "extra" ones then it should be fine?